moonsense/cli/download.py

```diff
@@ -0,0 +1,191 @@
+
+    localdir = os.getcwd()
+    datadir = os.path.join(localdir, "data")
+
+    while True:
+        if stop_event.is_set():
+            break
+
+        try:
+            msg = queue.get(timeout=.01)
+        except Empty:
+            # Run next iteration of loop
+            continue
+
+        if msg is None:
+            queue.task_done()
+
+        try:
+            session_id = msg
+            download_data_into_folder(datadir, with_group_id, session_id)
+        finally:
+            queue.task_done()
+
+
+def fetch_group_id(session):
+    return session.client_session_group_id
+
+
+def handler(signalname):
+    def wrapper_handler(signal_received, frame):
+        raise KeyboardInterrupt(f"{signalname} received")
+    return wrapper_handler
+
+def run_download(until: str, since: str, labels: list[str], with_group_id: bool = False):
+    """
+    Download sessions from a project.
+
+    :param until: Date in the YYYY-MM-DD format until the session data should be included.
+                  If not provided, the current day is used.
+    :param since: Date in the YYYY-MM-DD format since the session data should be included.
+                  If not provided beginning of Moonsense time - 1st of January 2021 is used.
+    :param labels: A list of labels to filter sessions by. A session needs to include at least
+                   one label in this list to be downloaded.
+    :param with_group_id: If set to True, organizes the downloaded sessions by date and
+                          client session group id. Default: False.
+    """
+    signal.signal(signal.SIGINT, handler("SIGINT"))
+    signal.signal(signal.SIGTERM, handler("SIGTERM"))
+
+    localdir = os.getcwd()
+    datadir = os.path.join(localdir, "data")
+    os.makedirs(datadir, exist_ok=True)
+
+    filter_by_since = datetime
+    if since is not None:
+        filter_by_since = datetime.strptime(since, "%Y-%m-%d").date()
+    else:
+        # beginning of Moonsense time is 1st of January 2021.
+        filter_by_since = datetime.strptime("2021-01-01", "%Y-%m-%d").date()
+
+    if until is not None:
+        filter_by_until = datetime.strptime(until, "%Y-%m-%d").date()
+    else:
+        filter_by_until = date.today()
+
+    queue = JoinableQueue()
+    stop_event = Event()
+    number_of_processes = os.cpu_count() * 2
+
+    all_procs = []
+    for process_count in range(number_of_processes):
+        local_process = Process(target=download_session,
+            daemon=True,
+            args=((queue, stop_event, with_group_id)))
+        all_procs.append(local_process)
+        local_process.start()
+
+    print("Downloading sessions in between", filter_by_since, filter_by_until)
+
+    session_counter = 0
+    filter_by_labels = labels if len(labels) > 0 else None
+
+    try:
+        # listing is in reverse chronological order - newest are first. 
+        for session in moonsense_client.list_sessions(filter_by_labels):        
```

Comment:
list sessions can take since/until as parameters. That would make it fairly efficient now that we use the index

Reply:
the backend supports it - the sdk needs to be updated to make the right request